File: typescript/tests/test_network.js
import assert from "node:assert/strict";
import { test } from "node:test";
import { Plane, SociaclError } from "../src/index.js";

function mesh() {
  const plane = new Plane();
  plane.addPerson("alice");
  plane.addPerson("bob");
  plane.addPerson("mallory");
  plane.addNetwork("panopticon");
  plane.addObject("panopticon", "alice");
  plane.setObjectProperty("panopticon", "predicate", "same-network");
  plane.admitMember("alice", "panopticon");
  plane.admitMember("bob", "panopticon");
  return plane;
}

test("same-network allows a member and denies an outsider", () => {
  const plane = mesh();
  assert.equal(plane.check("read", "panopticon", "bob", "same-network").allowed, true);
  assert.equal(plane.check("read", "panopticon", "mallory", "same-network").allowed, false);
  assert.equal(plane.remint("panopticon", "bob"), "remint");
});

test("one-sided in-network is not a grant", () => {
  const plane = new Plane();
  plane.addPerson("alice");
  plane.addPerson("bob");
  plane.addNetwork("mesh");
  plane.addObject("mesh", "alice");
  plane.setObjectProperty("mesh", "predicate", "same-network");
  plane.stateEdge("bob", "bob", "mesh", "in-network");
  assert.equal(plane.isMember("bob", "mesh"), false);
  assert.equal(plane.check("read", "mesh", "bob", "same-network").allowed, false);
});

test("owner censure for sabotage drops the member", () => {
  const plane = mesh();
  plane.censure("alice", "panopticon", "bob", "active-sabotage");
  assert.equal(plane.isMember("bob", "panopticon"), false);
  assert.equal(plane.audit("panopticon").length, 1);
});

test("member may self-leave", () => {
  const plane = mesh();
  plane.censure("bob", "panopticon", "bob", "self-leave");
  assert.equal(plane.isMember("bob", "panopticon"), false);
});

test("member cannot censure another", () => {
  const plane = mesh();
  assert.throws(
    () => plane.censure("bob", "panopticon", "alice", "policy-violation"),
    SociaclError,
  );
  assert.equal(plane.isMember("alice", "panopticon"), true);
});

test("owner cannot record self-leave for another", () => {
  const plane = mesh();
  assert.throws(
    () => plane.censure("alice", "panopticon", "bob", "self-leave"),
    SociaclError,
  );
  assert.equal(plane.isMember("bob", "panopticon"), true);
});

test("outsider cannot censure", () => {
  const plane = mesh();
  assert.throws(
    () => plane.censure("mallory", "panopticon", "bob", "active-sabotage"),
    SociaclError,
  );
});

File: python/tests/test_network.py
"""Network membership Python binding tests. Requires `cargo build -p sociacl-c`."""

from sociacl import Error, Plane


def _mesh() -> Plane:
    plane = Plane()
    plane.add_person("alice")
    plane.add_person("bob")
    plane.add_person("mallory")
    plane.add_network("panopticon")
    plane.add_object("panopticon", "alice")
    plane.set_object_property("panopticon", "predicate", "same-network")
    plane.admit_member("alice", "panopticon")
    plane.admit_member("bob", "panopticon")
    return plane


def test_same_network_membership():
    plane = _mesh()
    allowed, reason = plane.check("read", "panopticon", "bob", "same-network")
    assert allowed is True
    assert reason == "same-network"
    assert plane.is_member("bob", "panopticon") is True
    allowed, reason = plane.check("read", "panopticon", "mallory", "same-network")
    assert allowed is False
    assert plane.is_member("mallory", "panopticon") is False
    plane.close()


def test_owner_censure_drops_member():
    plane = _mesh()
    record = plane.censure("alice", "panopticon", "bob", "active-sabotage")
    assert "bob" in record
    assert "active-sabotage" in record
    allowed, _ = plane.check("read", "panopticon", "bob", "same-network")
    assert allowed is False
    assert plane.audit("panopticon")
    plane.close()


def test_member_cannot_censure_another():
    plane = _mesh()
    try:
        plane.censure("bob", "panopticon", "alice", "policy-violation")
        raise AssertionError("member must not censure another")
    except Error:
        pass
    assert plane.is_member("alice", "panopticon") is True
    plane.close()


def test_self_leave():
    plane = _mesh()
    plane.censure("bob", "panopticon", "bob", "self-leave")
    assert plane.is_member("bob", "panopticon") is False
    plane.close()


if __name__ == "__main__":
    test_same_network_membership()
    test_owner_censure_drops_member()
    test_member_cannot_censure_another()
    test_self_leave()
    print("ok")
